frontend/src/components/SignupForm.js

```diff
@@ -0,0 +1,126 @@
+import React, { useEffect, useState } from 'react';
+import { Link } from 'react-router-dom';
+import { useDispatch, useSelector } from 'react-redux';
+
+import { signup } from '../reducers/user';
+import { Wrapper, Form } from 'lib';
+
+const SignupForm = () => {
+  const dispatch = useDispatch();
+  const errorMessage = useSelector((store) => store.user.login.errorMessage);
+  const [name, setName] = useState('');
+  const [email, setEmail] = useState('');
+  const [password, setPassword] = useState('');
+
+  // This part can be improved with having one error object as a useState
+  const [nameError, setNameError] = useState(null);
+  const [emailError, setEmailError] = useState(null);
+  const [passwordError, setPasswordError] = useState(null);
+  const [showValidations, setShowValidations] = useState(false);
+  const [formIsValid, setFormIsValid] = useState(false);
+
+  const nameIsValid = () => {
+    if (!name) {
+      setNameError('Name can not be blank');
+      return false;
+    } else if (name.length < 2) {
+      setNameError('Name should be longer than 2 letters');
+      return false;
+    } else if (name.length > 40) {
+      setNameError('Name should be less than 40 letters');
+      return false;
+    }
+    setNameError(null);
+    return true;
+  };
+
+  const passwordIsValid = () => {
+    if (!password) {
+      setPasswordError('Please type in your password');
+      return false;
+    } else if (password.length < 5) {
+      setPasswordError('Password needs to be longer than 5 characters');
+      return false;
+    }
+    setPasswordError(null);
+    return true;
+  };
+
+  const emailIsValid = () => {
+    if (!email) {
+      setEmailError('Email can not be blank');
+      return false;
+    } else if (!email.includes('@')) {
+      setEmailError('Invalid email');
+      return false;
+    }
+    setEmailError(null);
+    return true;
+  };
+
+  useEffect(() => {
+    const formIsValid = nameIsValid() && passwordIsValid() && emailIsValid();
+    setFormIsValid(formIsValid);
+  }, [name, email, password]);
```

Comment:
Do you need useEffect to do this? Or could this be done in the handleSignUp function?
Also, does the useEffect run every time one of these are updated or when all 3 are updated? We're trying to understand this better.

Reply:
Let's see if I can try to explain it. As I see it YES we need it.

If you click Sig-Up with empty fields you will see that you get the frontend validation error message presented.

With the useEffect we have the possibility to remove/update the frontend validation error message if we change one of the variables name, email or password.